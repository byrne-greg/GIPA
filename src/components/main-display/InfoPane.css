.InfoPane__overlay {
  position: fixed; /* Sit on top of the page content */
  display: none; /* Hidden by default */
  width: 100%; /* Full width (cover the whole page) */
  height: 100%; /* Full height (cover the whole page) */
  top: 0; 
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #e5e5e5;
  z-index: 1; /* Specify a stack order */
  cursor: pointer; /* Add a pointer on hover */
  overflow-y: scroll;
}

.InfoPane__overlayOpenButton {
  border:none;
  margin:0 5px;
  color: #007fa5;
  background-color: white;
  font-size: calc(30px + 2vmin);
  cursor: pointer;
}

.InfoPane__content {
  margin: 5vh 5vw;
}

.Card__title {
  margin: 10px;
}

.Card__body {
  margin: 15px 30px;
  text-align: left;  
  font-size: calc(12px + 1.5vmin);
}

.Card__body p {
  padding: 15px 0;
  
}

.Card__image {
  display: block;
  margin-left: auto;
  margin-right: auto;
  height: auto;
  width: auto;
  max-width: 100%;
}

.Card__imagecaption {
  font-style: italic;
  text-align: center;
}

.CloseButton {
  position: fixed;
  bottom:0;
  right:0;
  height:50px;
  margin: 0 10px 0 0;
}

.CloseButton .fa-times-circle {
  font-size: calc(30px + 2vmin);
}