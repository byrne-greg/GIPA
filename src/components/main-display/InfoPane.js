import "./InfoPane.css";
import React from "react";
import PropTypes from "prop-types";

export default class InfoPane extends React.Component {
	constructor(props) {
		super(props);

		this.state = { isShowingInfoPaneOverlay: false };

		this.setIsShowingInfoPaneOverlay = this.setIsShowingInfoPaneOverlay.bind(
			this
		);
	}

	setIsShowingInfoPaneOverlay() {
		let prevState = this.state.isShowingInfoPaneOverlay;
		this.setState({ isShowingInfoPaneOverlay: !prevState });
	}

	determineDisplayStyle() {
		if (this.state.isShowingInfoPaneOverlay) {
			return "block";
		}
		return "none";
	}

	render() {
		return (
			<div className="InfoPane">
				<button
					className="InfoPane__overlayOpenButton"
					onClick={this.setIsShowingInfoPaneOverlay}
				>
					<i className="fas fa-info-circle" />
				</button>
				<div
					className="InfoPane__overlay"
					style={{ display: this.determineDisplayStyle() }}
				>
					<InfoPaneContent />
					<div
						className="InfoPane__suggestedResponse"
						onClick={this.setIsShowingInfoPaneOverlay}
					>
						{this.props.children}
					</div>
					<CloseButton closeOverlayFunc={this.setIsShowingInfoPaneOverlay} />
				</div>
			</div>
		);
	}
}

export const InfoPaneContent = props => {
	return (
		<div className="Card__content">
			<h2 className="Card__title">Greg's Interactive Personal Advocate</h2>
			<div className="Card__body">
				<p>
					GIPA was created as a personal project. The idea was to devise a way
					in which inquiring users could ask any questions they want to find out
					more about me. It's real mission was a self-learning practical
					exercise in using a wide variety of unfamiliar technologies and
					experimenting with my favourite topic at present time of writing -
					Chatbots!. Read about it's{" "}
					<a
						href="https://medium.com/@byrne.greg/a-development-story-in-creating-a-digital-advocate-with-watson-assistant-1623b824ec34"
						target="_blank"
						rel="noopener noreferrer"
					>
						development story
					</a>{" "}
					or visit the{" "}
					<a
						href="https://github.com/byrne-greg/GIPA"
						target="_blank"
						rel="noopener noreferrer"
					>
						code repository
					</a>
					.
				</p>
				<p>
					GIPA uses a bunch of technologies. The front-end was developed with{" "}
					<a
						href="https://reactjs.org/"
						target="_blank"
						rel="noopener noreferrer"
					>
						ReactJS
					</a>{" "}
					(and typical CSS/HTML). The conversational agent was created using the{" "}
					<a
						href="https://www.ibm.com/watson/ai-assistant/"
						target="_blank"
						rel="noopener noreferrer"
					>
						IBM's Watson Assistant
					</a>{" "}
					platform. The connectivity between the front-end and Watson Assistant
					uses the{" "}
					<a
						href="https://samnewman.io/patterns/architectural/bff/"
						target="_blank"
						rel="noopener noreferrer"
					>
						Backend for Frontend
					</a>{" "}
					pattern. It is deployed using a{" "}
					<a
						href="https://nodejs.org/en/"
						target="_blank"
						rel="noopener noreferrer"
					>
						NodeJS
					</a>{" "}
					FaaS with{" "}
					<a
						href="https://openwhisk.apache.org/"
						target="_blank"
						rel="noopener noreferrer"
					>
						Apache OpenWhisk
					</a>
					. The Backend for Frontend FaaS and the static front-end content is
					all hosted on{" "}
					<a
						href="https://www.ibm.com/cloud/"
						target="_blank"
						rel="noopener noreferrer"
					>
						IBM Cloud
					</a>{" "}
					(formerly Bluemix). The VCS uses{" "}
					<a
						href="https://github.com/"
						target="_blank"
						rel="noopener noreferrer"
					>
						GitHub
					</a>{" "}
					and the build toolchain uses{" "}
					<a
						href="https://travis-ci.org/"
						target="_blank"
						rel="noopener noreferrer"
					>
						Travis CI
					</a>{" "}
					and{" "}
					<a
						href="https://coveralls.io/"
						target="_blank"
						rel="noopener noreferrer"
					>
						Coveralls
					</a>
					.
				</p>
				<p>
					For functionality, the responses are hard-coded within the Assistant
					workspace (due to the mostly static information about me). There is
					some dynamic content such as asking for a joke, which is a call-out to{" "}
					<a
						href="https://icanhazdadjoke.com/"
						target="_blank"
						rel="noopener noreferrer"
					>
						https://icanhazdadjoke.com/
					</a>{" "}
					to retrieve a random joke.
				</p>
				<p>Try ask GIPA questions in natural language about me.</p>
				<div>{props.children}</div>
			</div>
		</div>
	);
};

export const CloseButton = props => {
	return (
		<div className="CloseButton" onClick={props.closeOverlayFunc}>
			<i className="far fa-times-circle" />
		</div>
	);
};
CloseButton.propTypes = {
	closeOverlayFunc: PropTypes.func.isRequired
};
