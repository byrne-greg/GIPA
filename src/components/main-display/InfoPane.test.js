import React from "react";
import InfoPane from "./InfoPane";
import renderer from "react-test-renderer";
import {shallow} from "enzyme/build";

describe("InfoPane component", () => {
  
	it("shallow renders without crashing", () => {
		shallow(<InfoPane />);
	});

	it("full render matches expected snapshot when not showing infopane", () => {
		const component = renderer.create(<InfoPane/>).toJSON();
		expect(component).toMatchSnapshot();
	});

	it("shallow renders with \"display: none\" by default", () => {
		const component = shallow(<InfoPane/>);
		expect(component.find(".InfoPane__overlay").prop("style").display).toEqual("none");
	});

	it("shallow renders with \"display: block\" when overlay button is clicked", () => {
		const component = shallow(<InfoPane/>);
		component.find(".InfoPane__overlayOpenButton").simulate("click");
		expect(component.find(".InfoPane__overlay").prop("style").display).toEqual("block");
	});

});