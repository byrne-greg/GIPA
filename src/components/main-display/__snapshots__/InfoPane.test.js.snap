// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoPane component full render matches expected snapshot when not showing infopane 1`] = `
<div
  className="InfoPane"
>
  <button
    className="InfoPane__overlayOpenButton"
    onClick={[Function]}
  >
    <i
      className="fas fa-info-circle"
    />
  </button>
  <div
    className="InfoPane__overlay"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="Card__content"
    >
      <h2
        className="Card__title"
      >
        Greg's Interactive Personal Advocate
      </h2>
      <div
        className="Card__body"
      >
        <p>
          GIPA was created as a personal project. The idea was to devise a way in which inquiring users could ask any questions they want to find out more about me. It's real mission was a self-learning practical exercise in using a wide variety of unfamiliar technologies and experimenting with my favourite topic at present time of writing - Chatbots! Read about it's
           
          <a
            href="https://medium.com/@byrne.greg/a-development-story-in-creating-a-digital-advocate-with-watson-assistant-1623b824ec34"
            rel="noopener noreferrer"
            target="_blank"
          >
            development story
          </a>
           
          or visit the
           
          <a
            href="https://github.com/byrne-greg/GIPA"
            rel="noopener noreferrer"
            target="_blank"
          >
            code repository
          </a>
          .
        </p>
        <p>
          GIPA uses a bunch of technologies. The front-end was developed with
           
          <a
            href="https://reactjs.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            ReactJS
          </a>
           
          (and typical CSS/HTML). The conversational agent was created using the
           
          <a
            href="https://www.ibm.com/watson/ai-assistant/"
            rel="noopener noreferrer"
            target="_blank"
          >
            IBM's Watson Assistant
          </a>
           
          platform. The connectivity between the front-end and Watson Assistant uses the
           
          <a
            href="https://samnewman.io/patterns/architectural/bff/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Backend for Frontend
          </a>
           
          pattern. It is deployed using a
           
          <a
            href="https://nodejs.org/en/"
            rel="noopener noreferrer"
            target="_blank"
          >
            NodeJS
          </a>
           
          FaaS with
           
          <a
            href="https://openwhisk.apache.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Apache OpenWhisk
          </a>
          . The Backend for Frontend FaaS and the static front-end content is all hosted on
           
          <a
            href="https://www.ibm.com/cloud/"
            rel="noopener noreferrer"
            target="_blank"
          >
            IBM Cloud
          </a>
           
          (formerly Bluemix) and Surge. The VCS uses
           
          <a
            href="https://github.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            GitHub
          </a>
           
          and the build toolchain uses
           
          <a
            href="https://travis-ci.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Travis CI
          </a>
           
          and
           
          <a
            href="https://coveralls.io/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Coveralls
          </a>
          .
        </p>
        <p>
          For functionality, the responses are hard-coded within the Assistant workspace (due to the mostly static information about me). There is some dynamic content such as asking for a joke, which is a call-out to
           
          <a
            href="https://icanhazdadjoke.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            https://icanhazdadjoke.com/
          </a>
           
          to retrieve a random joke.
        </p>
        <p>
          Try ask GIPA questions in natural language about me.
        </p>
        <div />
      </div>
    </div>
    <div
      className="InfoPane__suggestedResponse"
      onClick={[Function]}
    />
    <div
      className="CloseButton"
      onClick={[Function]}
    >
      <i
        className="far fa-times-circle"
      />
    </div>
  </div>
</div>
`;
